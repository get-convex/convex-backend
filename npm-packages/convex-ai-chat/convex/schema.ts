import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  messages: defineTable({
    isViewer: v.boolean(),
    sessionId: v.string(),
    text: v.string(),
  }).index("bySessionId", ["sessionId"]),
  documents: defineTable({
    url: v.string(),
    text: v.string(),
  }).index("byUrl", ["url"]),
  chunks: defineTable({
    documentId: v.id("documents"),
    text: v.string(),
    embeddingId: v.union(v.id("embeddings"), v.null()),
  })
    .index("byDocumentId", ["documentId"])
    .index("byEmbeddingId", ["embeddingId"]),
  embeddings: defineTable({
    embedding: v.array(v.number()),
    chunkId: v.id("chunks"),
  })
    .index("byChunkId", ["chunkId"])
    .vectorIndex("byEmbedding", {
      vectorField: "embedding",
      dimensions: 1536,
    }),
});
