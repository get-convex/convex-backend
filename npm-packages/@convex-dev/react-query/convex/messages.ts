import { mutation } from "./_generated/server.js";
import { query } from "./_generated/server.js";
import { Doc } from "./_generated/dataModel.js";

export const list = query(async (ctx): Promise<Doc<"messages">[]> => {
  return await ctx.db.query("messages").collect();
});

export const count = query(async (ctx): Promise<string> => {
  const messages = await ctx.db.query("messages").take(1001);
  return messages.length === 1001 ? "1000+" : `${messages.length}`;
});

export const send = mutation(
  async (ctx, { body, author }: { body: string; author: string }) => {
    const message = { body, author };
    await ctx.db.insert("messages", message);
  },
);
