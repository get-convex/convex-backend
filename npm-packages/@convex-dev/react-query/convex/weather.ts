import { action } from "./_generated/server.js";

export const getSFWeather = action(async () => {
  const stationUrl = "https://api.weather.gov/stations/SFOC1";
  const observationData = await (
    await fetch(`${stationUrl}/observations/latest`, {
      headers: { "User-Agent": "Convex/1.0" },
    })
  ).json();
  const celsius = observationData.properties.temperature.value;
  return {
    fahrenheit: Math.round(((celsius * 9) / 5 + 32) * 100) / 100,
    fetchedAt: Date.now(),
    timestamp: observationData.properties.timestamp,
  };
});
