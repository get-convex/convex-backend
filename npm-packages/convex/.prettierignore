temp
.git/
dist/
api-extractor-configs/reports
api-extractor-configs/temp
tmpDist*
management-openapi.json
src/cli/generatedApi.ts
