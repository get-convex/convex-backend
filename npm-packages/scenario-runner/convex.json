{
  "prodUrl": "http://127.0.0.1:8000",
  "project": "carnitas",
  "team": "carnitas",
  "functions": "convex/",
  "node": {
    "externalPackages": ["*"]
  }
}
