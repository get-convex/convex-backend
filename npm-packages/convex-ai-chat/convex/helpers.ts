import { PaginationResult } from "convex/server";
import { internal } from "./_generated/api";
import { Doc, TableNames } from "./_generated/dataModel";
import { ActionCtx, QueryCtx, internalQuery } from "./_generated/server";

export async function paginate<T extends TableNames>(
  ctx: ActionCtx,
  table: T,
  batchSize: number,
  callback: (documents: Doc<T>[]) => Promise<void>,
): Promise<void> {
  let isDone = false;
  let cursor = null;
  while (!isDone) {
    const result: PaginationResult<Doc<T>> = (await ctx.runQuery(
      internal.helpers.paginateQuery,
      {
        table,
        cursor,
        numItems: batchSize,
      },
    )) as any;
    await callback(result.page);
    ({ isDone, continueCursor: cursor } = result);
  }
}

export const paginateQuery = internalQuery(
  async <T extends TableNames>(
    ctx: QueryCtx,
    args: { table: T; cursor: any; numItems: number },
  ) => {
    return await ctx.db
      .query(args.table)
      .paginate({ cursor: args.cursor, numItems: args.numItems });
  },
);
