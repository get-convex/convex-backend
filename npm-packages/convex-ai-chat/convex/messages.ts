import { v } from "convex/values";
import { mutation } from "./_generated/server";
import { query } from "./_generated/server";
import { internal } from "./_generated/api";

export const list = query({
  args: {
    sessionId: v.string(),
  },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("messages")
      .withIndex("bySessionId", (q) => q.eq("sessionId", args.sessionId))
      .collect();
  },
});

export const send = mutation({
  args: {
    message: v.string(),
    sessionId: v.string(),
  },
  handler: async (ctx, { message, sessionId }) => {
    await ctx.db.insert("messages", {
      isViewer: true,
      text: message,
      sessionId,
    });
    await ctx.scheduler.runAfter(0, internal.serve.answer, {
      sessionId,
    });
  },
});
