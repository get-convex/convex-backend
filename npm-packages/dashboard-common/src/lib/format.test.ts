import { Shape } from "shapes";
import { ValidatorJSON } from "convex/values";
import {
  SchemaJson,
  displaySchema,
  displaySchemaFromShapes,
  prettier,
} from "./format";

type ShapeTypes = Shape["type"];

const id: Shape = {
  type: "Id",
  tableName: "test",
};

const shape = (type: ShapeTypes, more = {}): Shape =>
  // @ts-ignore
  ({ type, ...more });

const field = (name: string, fieldShape: Shape, optional: boolean = false) => ({
  fieldName: name,
  shape: fieldShape,
  optional,
});

const oneFieldTable = (fieldShape: Shape) =>
  shape("Object", {
    fields: [field("_id", id), field("field1", fieldShape)],
  });

const allPrimitiveShapes = shape("Object", {
  fields: [
    field("unknown", shape("Unknown")),
    field("never", shape("Never")),
    field("id", id),
    field("null", shape("Null")),
    field("bigint", shape("Int64")),
    field("number", shape("Float64")),
    field("boolean", shape("Boolean")),
    field("string", shape("String")),
    field("bytes", shape("Bytes")),
  ],
});

const object = shape("Object", {
  fields: [
    field(
      "object",
      shape("Object", {
        fields: [
          field("field1", shape("Boolean")),
          field("field1", shape("String")),
        ],
      }),
    ),
  ],
});

const array = shape("Array", {
  shape: shape("Boolean"),
});

const union = shape("Union", {
  shapes: [shape("Boolean"), shape("String"), shape("Bytes")],
});

const record = shape("Record", {
  keyShape: shape("String"),
  valueShape: {
    optional: false,
    shape: shape("Float64"),
  },
});

const complicatedNestedShape = shape("Object", {
  fields: [
    field(
      "field1",
      shape("Object", {
        fields: [
          field(
            "field1",
            shape("Array", {
              shape: shape("Union", {
                shapes: [shape("Boolean"), shape("String")],
              }),
            }),
          ),
        ],
      }),
    ),
    field(
      "field2",
      shape("Array", {
        shape: shape("Union", {
          shapes: [shape("Array", { shape: shape("String") }), shape("String")],
        }),
      }),
    ),
  ],
});

// System properties shouldn't appear in the generated schema (they are added automatically).
const shapeWithOnlySystemProperties = shape("Object", {
  fields: [field("_id", id), field("_creationTime", shape("Float64"))],
});

// If the shape contains a nested object with fields that match system properties
// they should appear in the generated schema.
const shapeWithNestedSystemProperties = shape("Object", {
  fields: [
    field(
      "nested",
      shape("Object", {
        fields: [field("_id", id), field("_creationTime", shape("Float64"))],
      }),
    ),
  ],
});

// In a top-level union, system fields should still not appear in the schema.
const shapeWithTopLevelUnion = shape("Union", {
  shapes: [
    shape("Object", {
      fields: [
        field("_id", id),
        field("_creationTime", shape("Float64")),
        field("left", shape("String")),
      ],
    }),
    shape("Object", {
      fields: [
        field("_id", id),
        field("_creationTime", shape("Float64")),
        field("right", shape("String")),
      ],
    }),
  ],
});

const shapeWithOptionalTopLevelField = shape("Object", {
  fields: [field("optionalString", shape("String"), true)],
});

const shapeWithOptionalNestedField = shape("Object", {
  fields: [
    field(
      "object",
      shape("Object", {
        fields: [field("optionalString", shape("String"), true)],
      }),
    ),
  ],
});

describe("displaySchemaFromShapes", () => {
  test.each([
    {
      name: "one boolean field",
      schema: new Map([["table1", oneFieldTable(shape("Boolean"))]]),
    },
    {
      name: "two tables",
      schema: new Map([
        ["table1", oneFieldTable(shape("Boolean"))],
        ["table2", oneFieldTable(shape("Boolean"))],
      ]),
    },
    {
      name: "no schema on all tables",
      schema: new Map([
        ["table1", shape("Never")],
        ["table2", shape("Never")],
      ]),
    },
    {
      name: "primitives shapes",
      schema: new Map([["table1", oneFieldTable(allPrimitiveShapes)]]),
    },
    {
      name: "object shape",
      schema: new Map([["table1", oneFieldTable(object)]]),
    },
    {
      name: "array shape",
      schema: new Map([["table1", oneFieldTable(array)]]),
    },
    {
      name: "union shape",
      schema: new Map([["table1", oneFieldTable(union)]]),
    },
    {
      name: "record shape",
      schema: new Map([["table1", oneFieldTable(record)]]),
    },
    {
      name: "complicated shape",
      schema: new Map([["table1", oneFieldTable(complicatedNestedShape)]]),
    },
    {
      name: "shape with only system properties",
      schema: new Map([["table1", shapeWithOnlySystemProperties]]),
    },
    {
      name: "shape with nested system properties",
      schema: new Map([["table1", shapeWithNestedSystemProperties]]),
    },
    {
      name: "shape with top level union",
      schema: new Map([["table1", shapeWithTopLevelUnion]]),
    },
    {
      name: "shape with optional top level field",
      schema: new Map([["table1", shapeWithOptionalTopLevelField]]),
    },
    {
      name: "shape with optional nested field",
      schema: new Map([["table1", shapeWithOptionalNestedField]]),
    },
  ])("$name", ({ schema }) => {
    expect(displaySchemaFromShapes(schema)).toMatchSnapshot();
  });

  test("displaySchema with only search indexes", () => {
    const schemaJson: SchemaJson = {
      tables: [
        {
          tableName: "table",
          indexes: [],
          searchIndexes: [
            {
              indexDescriptor: "search_index",
              searchField: "property1",
              filterFields: [],
            },
          ],
          vectorIndexes: [],
          documentType: {
            type: "any",
          },
        },
        {
          tableName: "table_next",
          indexes: [],
          searchIndexes: [],
          vectorIndexes: [],
          documentType: {
            type: "any",
          },
        },
      ],
      schemaValidation: true,
    };
    expect(displaySchema(schemaJson)).toMatchSnapshot();
  });

  test("displaySchema with only vector indexes", () => {
    const schemaJson: SchemaJson = {
      tables: [
        {
          tableName: "table",
          indexes: [],
          searchIndexes: [],
          vectorIndexes: [
            {
              indexDescriptor: "vector_index",
              vectorField: "property1",
              dimensions: 1536,
              filterFields: [],
            },
          ],
          documentType: {
            type: "any",
          },
        },
        {
          tableName: "table_next",
          indexes: [],
          searchIndexes: [],
          vectorIndexes: [],
          documentType: {
            type: "any",
          },
        },
      ],
      schemaValidation: true,
    };
    expect(displaySchema(schemaJson)).toMatchSnapshot();
  });

  // Older schemas will not have a vector index field.
  test("displaySchema without vector indexes", () => {
    const schemaJson: SchemaJson = {
      tables: [
        {
          tableName: "table",
          indexes: [],
          searchIndexes: [],
          documentType: {
            type: "any",
          },
        } as any,
      ],
      schemaValidation: true,
    };
    expect(displaySchema(schemaJson)).toMatchSnapshot();
  });

  test("displaySchema", () => {
    const documentType: ValidatorJSON = {
      type: "object",
      value: {
        ref: {
          fieldType: { type: "id", tableName: "reference" },
          optional: false,
        },
        nullField: { fieldType: { type: "null" }, optional: false },
        numberField: { fieldType: { type: "number" }, optional: false },
        bigintField: { fieldType: { type: "bigint" }, optional: false },
        booleanField: { fieldType: { type: "boolean" }, optional: false },
        stringField: { fieldType: { type: "string" }, optional: false },
        bytesField: { fieldType: { type: "bytes" }, optional: false },
        arrayField: {
          fieldType: { type: "array", value: { type: "boolean" } },
          optional: false,
        },
        anyField: { fieldType: { type: "any" }, optional: false },
        literalBigint: {
          fieldType: {
            type: "literal",
            value: {
              $integer: "AQAAAAAAAAA=",
            },
          },
          optional: false,
        },
        literalNumber: {
          fieldType: {
            type: "literal",
            value: 0.0,
          },
          optional: false,
        },
        literalString: {
          fieldType: {
            type: "literal",
            value: "hello world",
          },
          optional: false,
        },
        literalBoolean: {
          fieldType: {
            type: "literal",
            value: true,
          },
          optional: false,
        },
        union: {
          fieldType: {
            type: "union",
            value: [{ type: "string" }, { type: "number" }],
          },
          optional: false,
        },
        object: {
          fieldType: {
            type: "object",
            value: {
              a: { fieldType: { type: "any" }, optional: true },
            },
          },
          optional: false,
        },
      },
    };
    const schemaJson: SchemaJson = {
      tables: [
        {
          tableName: "table",
          indexes: [
            { indexDescriptor: "by_a", fields: ["a"] },
            { indexDescriptor: "by_a_b", fields: ["a", "b"] },
          ],
          searchIndexes: [
            {
              indexDescriptor: "no_filter_fields",
              searchField: "property1",
              filterFields: [],
            },
            {
              indexDescriptor: "one_filter_field",
              searchField: "property1",
              filterFields: ["property1"],
            },
            {
              indexDescriptor: "two_filter_fields",
              searchField: "property1",
              filterFields: ["property1", "property2"],
            },
          ],
          vectorIndexes: [
            {
              indexDescriptor: "vector_no_filter_fields",
              vectorField: "property1",
              dimensions: 1536,
              filterFields: [],
            },
            {
              indexDescriptor: "vector_one_filter_field",
              vectorField: "property1",
              dimensions: 1536,
              filterFields: ["property1"],
            },
            {
              indexDescriptor: "vector_two_filter_fields",
              vectorField: "property1",
              dimensions: 1536,
              filterFields: ["property1", "property2"],
            },
          ],
          documentType,
        },
        {
          tableName: "table_any",
          indexes: [],
          searchIndexes: [],
          vectorIndexes: [],
          documentType: {
            type: "any",
          },
        },
        {
          tableName: "table_union",
          indexes: [],
          searchIndexes: [],
          vectorIndexes: [],
          documentType: {
            type: "union",
            value: [
              {
                type: "object",
                value: {
                  a: { fieldType: { type: "any" }, optional: false },
                  z: { fieldType: { type: "any" }, optional: true },
                },
              },
              {
                type: "object",
                value: {
                  b: { fieldType: { type: "any" }, optional: true },
                },
              },
            ],
          },
        },
      ],
      schemaValidation: true,
    };
    expect(displaySchema(schemaJson)).toMatchSnapshot();
  });
  test("schema validation false", () => {
    const schemaJson = { tables: [], schemaValidation: false };
    expect(displaySchema(schemaJson)).toMatchSnapshot();
  });
});

describe("prettier", () => {
  test("format succeeds", () => {
    const result = prettier("const a = 1;");
    expect(result).toBe("const a = 1;");
  });

  test("format does not include source code", () => {
    // This error for this code with a syntax error should not include the syntax error.
    const fn = () => {
      prettier("const a = ;");
    };
    expect(fn).toThrowError("Unexpected token (1:11)");
  });
});
