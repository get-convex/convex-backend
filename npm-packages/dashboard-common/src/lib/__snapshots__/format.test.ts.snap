// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displaySchemaFromShapes array shape 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({ field1: v.array(v.boolean()) }),
});"
`;

exports[`displaySchemaFromShapes complicated shape 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    field1: v.object({
      field1: v.object({
        field1: v.array(v.union(v.boolean(), v.string())),
      }),
      field2: v.array(
        v.union(v.array(v.string()), v.string())
      ),
    }),
  }),
});"
`;

exports[`displaySchemaFromShapes displaySchema 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table: defineTable({
    ref: v.id("reference"),
    nullField: v.null(),
    numberField: v.float64(),
    bigintField: v.int64(),
    booleanField: v.boolean(),
    stringField: v.string(),
    bytesField: v.bytes(),
    arrayField: v.array(v.boolean()),
    anyField: v.any(),
    literalBigint: v.literal(1n),
    literalNumber: v.literal(0),
    literalString: v.literal("hello world"),
    literalBoolean: v.literal(true),
    union: v.union(v.string(), v.float64()),
    object: v.object({ a: v.optional(v.any()) }),
  })
    .index("by_a", ["a"])
    .index("by_a_b", ["a", "b"])
    .searchIndex("no_filter_fields", {
      searchField: "property1",
    })
    .searchIndex("one_filter_field", {
      searchField: "property1",
      filterFields: ["property1"],
    })
    .searchIndex("two_filter_fields", {
      searchField: "property1",
      filterFields: ["property1", "property2"],
    })
    .vectorIndex("vector_no_filter_fields", {
      vectorField: "property1",
      dimensions: 1536,
    })
    .vectorIndex("vector_one_filter_field", {
      vectorField: "property1",
      dimensions: 1536,
      filterFields: ["property1"],
    })
    .vectorIndex("vector_two_filter_fields", {
      vectorField: "property1",
      dimensions: 1536,
      filterFields: ["property1", "property2"],
    }),
  table_any: defineTable(v.any()),
  table_union: defineTable(
    v.union(
      v.object({ a: v.any(), z: v.optional(v.any()) }),
      v.object({ b: v.optional(v.any()) })
    )
  ),
});"
`;

exports[`displaySchemaFromShapes displaySchema with only search indexes 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table: defineTable(v.any()).searchIndex("search_index", {
    searchField: "property1",
  }),
  table_next: defineTable(v.any()),
});"
`;

exports[`displaySchemaFromShapes displaySchema with only vector indexes 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table: defineTable(v.any()).vectorIndex("vector_index", {
    vectorField: "property1",
    dimensions: 1536,
  }),
  table_next: defineTable(v.any()),
});"
`;

exports[`displaySchemaFromShapes displaySchema without vector indexes 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table: defineTable(v.any()),
});"
`;

exports[`displaySchemaFromShapes no schema on all tables 1`] = `undefined`;

exports[`displaySchemaFromShapes object shape 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    field1: v.object({
      object: v.object({
        field1: v.boolean(),
        field1: v.string(),
      }),
    }),
  }),
});"
`;

exports[`displaySchemaFromShapes one boolean field 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({ field1: v.boolean() }),
});"
`;

exports[`displaySchemaFromShapes primitives shapes 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    field1: v.object({
      unknown: v.any(),
      never: v.any(),
      id: v.id("test"),
      null: v.null(),
      bigint: v.int64(),
      number: v.float64(),
      boolean: v.boolean(),
      string: v.string(),
      bytes: v.bytes(),
    }),
  }),
});"
`;

exports[`displaySchemaFromShapes record shape 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    field1: v.record(v.string(), v.float64()),
  }),
});"
`;

exports[`displaySchemaFromShapes schema validation false 1`] = `undefined`;

exports[`displaySchemaFromShapes shape with nested system properties 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    nested: v.object({
      _id: v.id("test"),
      _creationTime: v.float64(),
    }),
  }),
});"
`;

exports[`displaySchemaFromShapes shape with only system properties 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({ table1: defineTable({}) });"
`;

exports[`displaySchemaFromShapes shape with optional nested field 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    object: v.object({
      optionalString: v.optional(v.string()),
    }),
  }),
});"
`;

exports[`displaySchemaFromShapes shape with optional top level field 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    optionalString: v.optional(v.string()),
  }),
});"
`;

exports[`displaySchemaFromShapes shape with top level union 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable(
    v.union(
      v.object({ left: v.string() }),
      v.object({ right: v.string() })
    )
  ),
});"
`;

exports[`displaySchemaFromShapes two tables 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({ field1: v.boolean() }),
  table2: defineTable({ field1: v.boolean() }),
});"
`;

exports[`displaySchemaFromShapes union shape 1`] = `
"import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  table1: defineTable({
    field1: v.union(v.boolean(), v.string(), v.bytes()),
  }),
});"
`;
